expression/builtin_compare.go

```diff
@@ -82,40 +88,190 @@ var (
 	_ builtinFunc = &builtinNullEQDurationSig{}
 )
 
+// Coalesce returns the first non-NULL value in the list,
+// or NULL if there are no non-NULL values.
 type coalesceFunctionClass struct {
 	baseFunctionClass
 }
 
-func (c *coalesceFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	if err := c.verifyArgs(args); err != nil {
+func (c *coalesceFunctionClass) getFunction(args []Expression, ctx context.Context) (sig builtinFunc, err error) {
+	if err = c.verifyArgs(args); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	fieldTps := make([]*types.FieldType, 0, len(args))
+	for _, arg := range args {
+		fieldTps = append(fieldTps, arg.GetType())
+	}
+
+	// Get the aggregate field type as retType
+	retTp := types.AggFieldType(fieldTps)
+	retCTp := types.AggTypeClass(fieldTps, &retTp.Flag)
+	retEvalTp := fieldTp2EvalTp(retTp)
+
+	fieldEvalTps := make([]evalTp, 0, len(args))
+	for range args {
+		fieldEvalTps = append(fieldEvalTps, retEvalTp)
+	}
+
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, retEvalTp, fieldEvalTps...)
+	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	sig := &builtinCoalesceSig{newBaseBuiltinFunc(args, ctx)}
+
+	bf.tp.Flag |= retTp.Flag
+	// Set retType to BINARY(0) if all arguments are of type NULL
+	retTp.Flen, retTp.Decimal = 0, types.UnspecifiedLength
+	if retTp.Tp == mysql.TypeNull {
+		types.SetBinChsClnFlag(bf.tp)
+	} else {
+		maxIntLen := 0
```

Comment:
this code is not easy to understand
may add a comment for this infering process?